src/lib/operators/index_scan.cpp

```diff
@@ -0,0 +1,154 @@
+#include "index_scan.hpp"
+
+#include <algorithm>
+
+#include "scheduler/abstract_task.hpp"
+#include "scheduler/current_scheduler.hpp"
+#include "scheduler/job_task.hpp"
+
+#include "storage/index/base_index.hpp"
+#include "storage/reference_column.hpp"
+
+#include "utils/assert.hpp"
+
+
+namespace opossum {
+
+IndexScan::IndexScan(std::shared_ptr<AbstractOperator> in, std::vector<ChunkID> chunk_ids,
+                     const ColumnIndexType index_type, std::vector<ColumnID> left_column_ids, const ScanType scan_type,
+                     std::vector<AllTypeVariant> right_values, std::vector<AllTypeVariant> right_values2)
+    : AbstractReadOnlyOperator{in},
+      _chunk_ids{chunk_ids},
+      _index_type{index_type},
+      _left_column_ids{left_column_ids},
+      _scan_type{scan_type},
+      _right_values{right_values},
+      _right_values2{right_values2} {}
+
+const std::string IndexScan::name() const { return "IndexScan"; }
+
+std::shared_ptr<const Table> IndexScan::_on_execute() {
+  _in_table = _input_table_left();
+
+  _validate_input();
+
+  _out_table = Table::create_with_layout_from(_in_table);
+
+  std::mutex output_mutex;
+
+  auto jobs = std::vector<std::shared_ptr<AbstractTask>>{};
+  jobs.reserve(_chunk_ids.size());
+
+  for (auto chunk_id : _chunk_ids) {
+    auto job_task = std::make_shared<JobTask>([=, &output_mutex]() {
+      const auto chunk_guard = _in_table->get_chunk_with_access_counting(chunk_id);
```

Comment:
Thanks for thinking about NUMA! Please reduce the scope so that the time spent when building the pos list and trying to acquire the lock is not included.

Reply:
I moved chunk_guard under the creation of the pos_list. I copied this code. I’m not sure what it actually does. In which cases should this guard be used?

Reply:
It's not really a guard. It tracks how long the chunk was used in order to estimate the number of accesses.

Reply:
You should hold it for as long as you actually touch the chunk itself.

Reply:
Including its columns?

Reply:
Hmm. Indexes are not even part of it yet. I think it would be safer to keep it out for now.